paddle/framework/lod_tensor.h

```diff
@@ -0,0 +1,226 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#pragma once
+
+#include <memory>
+#if (!PADDLE_ONLY_CPU)
+#include <thrust/device_vector.h>
+#include <thrust/host_vector.h>
+#endif
+
+#include "paddle/framework/ddim.h"
+#include "paddle/framework/tensor.h"
+#include "paddle/platform/enforce.h"
+
+namespace paddle {
+namespace framework {
+
+/*
+ * LODTensor (Level of details Tensor)
+ * see https://en.wikipedia.org/wiki/Level_of_details for reference.
+ */
+class LODTensor {
+ public:
+// level_t save offsets of each unit.
+#ifdef PADDLE_ONLY_CPU
+  using level_t = std::vector<size_t>;
+#else
+  using level_t = thrust::device_vector<size_t>;
+#endif
+  // LOD stores offsets of each level of units, the largest units level first,
+  // then the smaller units level. Each level_t stores the offsets of units in
+  // Tesor.
+  typedef std::vector<level_t> lod_t;
+
+  LODTensor() {}
+  LODTensor(std::shared_ptr<Tensor> tensor, std::shared_ptr<lod_t> lod) {
+    Reset(tensor, lod);
+  }
+
+  void Reset(std::shared_ptr<Tensor> tensor, std::shared_ptr<lod_t> lod) {
+    tensor_ = tensor;
+    lod_start_pos_ = lod;
+  }
+
+  /*
+   * Get a element from LOD.
+   */
+  size_t lod_element(size_t level, size_t elem) const {
+    PADDLE_ENFORCE(level < Levels(), "level [%d] out of range [%d]", level,
+                   Levels());
+    PADDLE_ENFORCE(elem < Elements(level),
+                   "element begin [%d] out of range [%d]", elem,
+                   Elements(level));
+    return lod_start_pos_->at(level)[elem];
```

Comment:
`lod_start_pos_[level][elem];` is better, because the previous `PADDLE_ENFORCE` has check the index of `lod_start_pos_`, it is not necessary use `at` to check it again.

The different between `at` and `[]` is at will perform a range check, and throw a `std::runtime_error` when out of range.